bridge/src/reactor.rs

```diff
@@ -0,0 +1,124 @@
+use crate::connection::{Connection, ConnectionError};
+use tokio::runtime::Builder;
+use tokio::sync::{mpsc, oneshot};
+
+#[derive(Debug)]
+pub enum Request {
+    Connect {
+        host: String
+    },
+    Rpc {
+        connection: Connection,
+        method: String,
+        bytes: Vec<u8>
+    }
+}
+
+#[derive(Debug)]
+pub enum Response {
+    Empty {},
+    Connection {
+        connection: Connection
+    },
+    Rpc {
+        bytes: Vec<u8>
+    }
+}
+
+#[derive(Debug)]
+pub enum ReactorError {
+    ConnectionError(ConnectionError)
+}
+
+impl std::fmt::Display for ReactorError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ReactorError::ConnectionError(error) => write!(f, "{}", error),
+        }
+    }
+}
+
+impl From<ConnectionError> for ReactorError {
+    fn from(err: ConnectionError) -> Self {
+        ReactorError::ConnectionError(err)
+    }
+}
+
+impl std::error::Error for ReactorError {}
+
+pub type ReactorResult = Result<Response, ReactorError>;
+
+// A request wrapper which includes a channel for sending back the response
+#[derive(Debug)]
+struct Envelope {
+    request: Request,
+    channel: oneshot::Sender<ReactorResult>
+}
+
+async fn handle_request(request: Request, channel: oneshot::Sender<ReactorResult>) {
+    match request {
+        Request::Connect { host } => {
+            let result = Connection::connect(host).await
+                .map(|connection| { Response::Connection { connection: connection } })
+                .map_err(|error| ReactorError::ConnectionError(error));
+
+            channel.send(result);
+        },
+        Request::Rpc { mut connection, method, bytes } => {
+            let result = connection.call(&method, bytes).await
+                .map(|bytes| { Response::Rpc { bytes: bytes } })
+                .map_err(|error| ReactorError::ConnectionError(error));
+
+            channel.send(result);
+        }
+    }
+}
+
+fn start_loop(thread_count: usize, receiver: &mut mpsc::UnboundedReceiver<Envelope>) {
+    let runtime = Builder::new_multi_thread()
+        .worker_threads(thread_count)
+        .enable_all()
+        .thread_name("core")
+        .build()
+        .unwrap();
+
+    runtime.block_on(async move {
+        while let Some(msg) = receiver.recv().await {
+            tokio::spawn(handle_request(msg.request, msg.channel));
+        }
+    });
+}
+
+// A Reactor is responsible for processing requests asynchronously using an event loop
+// while exposing a synchronous interface to the callers
+
+#[derive(Clone)]
+pub struct Reactor {
+    sender: mpsc::UnboundedSender<Envelope>,
```

Comment:
Since you're gonna eat up a thread waiting for response anyways, what value does a channel buy you over just using tokio's task management? I think it'd be much clearer if you stored the tokio runtime right here in the reactor and did literal `tokio::spawn` calls right where you need them (e.g. in `process`, though you don't even really need a generic `process` call then, you can just have a normal rpc call).